---
title: "circRNA pipeline README"
author: "Xianjun Dong, PhD"
date: "October 27, 2016; May 30, 2017"
output: html_document
---

## Introduction

This is a short summary for the pipeline of circRNA analysis, including:

### Paper1: Circular RNAs in human brain neurons of neurodegenerative diseases
1. circRNA detection in >200 brain samples (SNDA: 117; TCPY: 94; MCPY: 3; NN: 8)
2. cell-type specific expression of circRNAs (SNDA, TCPY, FB, PBMC)
3. PD (N=22) vs. ILB (N=28) vc. HC (N=67) SNDA circRNAs  --> validated in CFS (PD vs. HC)
4. AD (N=44) vc. HC (N=50) TCPY circRNAs --> detected in CSF (HC)

### Paper2: Genetic effects on the expression, splicing and circularization of circRNAs
4. eQTL analysis of circRNAs in SNDA samples
5. allele-specific circularization
6. disease/trait GWAS enrichment

## 1. circRNA annotation
by combining the work from two rotation students Rebeca and Alyssa

### Rebeca's part 

ran circExplorer on the 12 new samples (with RNaseR and mock, by Yunfei) and all BRAINCODE healthy samples, and merge together

**1. ran circExplorer on each sample**

`RNAseq.pipeline.sh` is the final code for RNase R and Mock samples (just circExplorer and not Mapsplice).

Note from Rebeca: "`RNAseq.pipeline.mapsplice221.Braincode.sh` and `RNAseq.pipeline.mapsplice.sh` were the most finalized versions of the mapsplice code for Braincode and the RNaseR samples, but I was still getting many errors after days of running mapsplice. Mostly in files with high sequencing depth. I then decided to test either changing some parameters to make it more efficient (`RNAseq.pipeline.mapsplice221_canonical.sh`) or to subsample the files (`RNAseq.pipeline.mapsplice221.subsample.sh`) but I never got that to work and focused on analyzing only circexplorer results since those did run successfully for all samples. So the code I eventually used is `RNAseq.pipeline.sh` which only runs circexplorer." (from 20161027 email)

Note: Rerun all samples using the modified version of gene reference (+CDR1as) (2017/10/28):
For RM samples: $HOME/neurogen/circRNA_seq_Rebeca_HC/Code_Executables/pipeline/RNAseq/RNAseq.pipeline.sh /data/neurogen/circRNA_seq_Rebeca_HC/rawfiles  (2018/11/13: rerun with CIRCexploer2. Actually it's same as circ.txt from CIRCexploer.)
For BC samples: $HOME/neurogen/pipeline/RNAseq/RNAseq.pipeline.sh /data/neurogen/rnaseq_PD/rawfiles

**2. merge circRNA called from all RM samples**

merge CircExplorer output files "circularRNA_known.txt" to count circRNAs for several samples in one table.  

```{bash}
## all RM samples by Yunfei
cd /data/neurogen/circRNA_seq_Rebeca_HC/
mkdir RM_circexp
cd RM_circexp
# only include HCLIB and the first runs (not those that were resequenced) 
echo "HC_WGC082362_SN_R_b1_r1
HC_WGC082363_SN_M_b1_r1
HC_WGC082364_SN_R_b2_r1
HC_WGC082365_SN_M_b2_r1
HC_MD6326_MB_M_b3
HC_MD6326_MB_R_b3
HC_WGC082366_TC_R_b1
HC_WGC082367_TC_M_b1
HC_WGC082368_TC_R_b2
HC_WGC082369_TC_M_b2
HC_WGC082370_FB_R_b1
HC_WGC082371_FB_M_b1
HC_WGC082372_PBMC_R_b1
HC_WGC082373_PBMC_M_b1" > RM.samplelist 
ls ~/neurogen/circRNA_seq_Rebeca_HC/run_output/*/circularRNA_known.txt | grep -f RM.samplelist - | while read line; do i=${line/*run_output\//}; ii=${i/\/circularRNA_known.txt/_candidates.bed}; echo $ii; ln -fs $line $ii; done
module load python/2.7.3
python /data/neurogen/circRNA_seq_Rebeca_HC/Code_Executables/Merge_circRNA_tools/merge_circexplorer.py *candidates.bed | sed 's/WGC082362/MGH1026/;s/WGC082363/MGH1026/;s/WGC082364/MC3290/;s/WGC082365/MC3290/;s/WGC082366/TCKY1247/;s/WGC082367/TCKY1247/;s/WGC082368/TCKY1217/;s/WGC082369/TCKY1217/;s/WGC082370/ND34770/;s/WGC082371/ND34770/;s/WGC082372/H02018/;s/WGC082373/H02018/' > Merge_circexplorer_RM.rawcount.txt
echo -e "chrom\tstart\tend\tID\tscore\tstrand\tthickStart\tthickEnd\titemRgb\texonCount\texonSizes\texonOffsets\tcircType\tgeneName" > Merge_circexplorer_RM.annotation.bed14
awk -v OFS='\t' '{print $1,$2,$3,$1"_"$2"_"$3,1,$6,$7,$8,$9,$10,$11,$12,($14=="Yes" || $14=="ciRNA")?"ciRNA":"circRNA",$15}' *_candidates.bed | sort -u >>  Merge_circexplorer_RM.annotation.bed14

```

**3. merge circRNA called from all BC samples**

```{bash}
## HC only
# cd /data/neurogen/circRNA_seq_Rebeca_HC/
# mkdir BRAINCODE_circexp
# cd BRAINCODE_circexp
# for subdir in ~/neurogen/rnaseq_PD/run_output/HC*; do cp $subdir/circ.txt ${subdir}_candidates.bed; done;
# mkdir NotInclude
# mv HC*{unamplified,strand}*candidates.bed  NotInclude/
# module load python/2.7.3
# python /data/neurogen/circRNA_seq_Rebeca_HC/Code_Executables/Merge_circRNA_tools/merge_circexplorer.py *candidates.bed > Merge_circexplorer_BC.bed
# cat Merge_circexplorer_BC.bed| cut -f1,2,3,6 > Merge_coords_BC.bed
# awk -v OFS='\t' '{print $1,$2,$3,$6,$10,$11,$14,$15}' *_candidates.bed | sort -k1,1 -k2,2 | uniq  >  concatenated_circOut_unique_BC.txt

## change to use all 106 samples (HC+ILB) in Fig. S1a
cd /data/neurogen/circRNA_seq_Rebeca_HC/
#mv BRAINCODE_circexp _BRAINCODE_circexp_rebeca; mkdir BRAINCODE_circexp  # backup of Rebeca's original result 
cd /data/neurogen/circRNA_seq_Rebeca_HC/BRAINCODE_circexp
cat ~/neurogen/rnaseq_PD/results/merged/samplelist.HCILB_SNDA ~/neurogen/rnaseq_PD/results/merged/samplelist.HC_PY ~/neurogen/rnaseq_PD/results/merged/samplelist.HC_nonNeuron ~/neurogen/rnaseq_PD/results/merged/samplelist.PD_SNDA > BC.n125.samplelist
cat ~/neurogen/rnaseq_PD/results/merged/samplelist.HCILB_SNDA ~/neurogen/rnaseq_PD/results/merged/samplelist.HC_PY ~/neurogen/rnaseq_PD/results/merged/samplelist.HC_nonNeuron > BC.n106.samplelist
# update to use the output from CIRCexploer2
ls ~/neurogen/rnaseq_PD/run_output/*/circularRNA_known.txt | grep -f BC.n125.samplelist - | while read line; do i=${line/*run_output\//}; ii=${i/\/circularRNA_known.txt/_candidates.bed}; echo $ii; ln -fs $line $ii; done
module load python/2.7.3
python /data/neurogen/circRNA_seq_Rebeca_HC/Code_Executables/Merge_circRNA_tools/merge_circexplorer.py *candidates.bed > Merge_circexplorer_BC.rawcount.txt
echo -e "chrom\tstart\tend\tID\tscore\tstrand\tthickStart\tthickEnd\titemRgb\texonCount\texonSizes\texonOffsets\tcircType\tgeneName" > Merge_circexplorer_BC.annotation.bed14
awk -v OFS='\t' '{print $1,$2,$3,$1"_"$2"_"$3,1,$6,$7,$8,$9,$10,$11,$12,($14=="Yes" || $14=="ciRNA")?"ciRNA":"circRNA",$15}' *_candidates.bed | sort -u >>  Merge_circexplorer_BC.annotation.bed14
```


**4. merge circRNA called from RM and BC samples [optional] **

```{bash}
cd ~/projects/circRNA/data
ln -fs ~/neurogen/circRNA_seq_Rebeca_HC/BRAINCODE_circexp/Merge_circexplorer_BC.annotation.bed14 Merge_circexplorer_BC.annotation.bed14
ln -fs ~/neurogen/circRNA_seq_Rebeca_HC/BRAINCODE_circexp/Merge_circexplorer_BC.rawcount.txt Merge_circexplorer_BC.rawcount.txt
ln -fs ~/neurogen/circRNA_seq_Rebeca_HC/RM_circexp/Merge_circexplorer_RM.annotation.bed14 Merge_circexplorer_RM.annotation.bed14
ln -fs ~/neurogen/circRNA_seq_Rebeca_HC/RM_circexp/Merge_circexplorer_RM.rawcount.txt Merge_circexplorer_RM.rawcount.txt

cat Merge_circexplorer_BC.annotation.bed14 Merge_circexplorer_RM.annotation.bed14 | grep -v chrom | sort -k1,1 -k2,2n -u  >  Merge_circexplorer_BC_RM.annotation.bed14
cut -f1,2,3,6,10,11,13,14 Merge_circexplorer_BC_RM.annotation.bed14 > concatenated_circOut_unique_BC_RM.txt

## track for UCSC 
awk '{OFS="\t";$9=($13=="circRNA")?(($6=="+")?"255,0,0":"0,0,255"):(($6=="+")?"255,100,100":"100,100,255");print}' Merge_circexplorer_BC_RM.annotation.bed14 | cut -f1-12 >  Merge_circexplorer_BC_RM.annotation.bed12
bedToBigBed -type=bed12 Merge_circexplorer_BC_RM.annotation.bed12 http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/hg19.chrom.sizes Merge_circexplorer_BC_RM.annotation.bb
chmod 644 Merge_circexplorer_BC_RM.annotation.bb;
scp Merge_circexplorer_BC_RM.annotation.bb xd010@panda.dipr.partners.org:~/public_html/tracks
# add the following description to http://panda.partners.org/~xd010/myHub/hg19/trackDb.circRNA.txt
    track BC_BM_circRNAs
    bigDataUrl http://pd:brain@panda.partners.org/~xd010/tracks/Merge_circexplorer_BC_RM.annotation.bb
    shortLabel BC_RM_circularRNA
    longLabel Circular RNAs from all RM and BC samples (see README.txt in src)
    visibility pack
    itemRgb on
    type bigBed 12
    parent circRNA_braincode

```

**5. total table, annotaton, and fold change**

```{bash}
cd ~/projects/circRNA/data
# run the third part of ~/projects/circRNA/data/Files_Rebeca_borges/Circ_rnas_noOutliers.update.R or equally makeTotalTable.R
Rscript makeTotalTable.R
```
It will generate Table_annotation.txt, Table_expression.txt, and all fold-change files xx_BC_R_M_union_FoldChange.txt for SN, TC, PBMC, and FB, also sorted annotation according to mean RNase values in each cell type [for Yunfei]

Note that we remove one SN sample as its RNase experiment does not enrich circRNA (see detail in the code comment)

### Alyssa's part 

run CIRI on the BRAINCODE samples, etc. [TO BE ADDED]


## 2. cell-type specific expression of circRNAs
Definition of "being expressed": 

- with >=2 unique back-splicing reads in at least 1 sample (Zheng et al., doi:10.1038/ncomms11215)

Definition of "being cell-type specific": 

- in the specific tissue, specificity score S>=0.5 AND mean expression > mean+2s.d. of overall expression (Zheng et al., doi:10.1038/ncomms11215)

## 3. eQTL analysis of circRNAs in SNDA samples ##

- input high-confidence circRNAs called from above (n=10017)
- filter with expression of >0.001 RPM in at least 5% individuals and â‰¥1 reads in at least 5% individuals  (n=601)

_SVA.eQTL.R or fastqQTL.main.sh ?



## 4. allele-specific circularization
## 5. disease/trait GWAS enrichment